README.md
**N.B. The Bellroy technology team are transitioning away from Ruby as their primary programming language, and will not longer actively update this repository. Contact us if you'd like to take over as the primary maintainer.**

![dynamic-rendering-logo](https://user-images.githubusercontent.com/2643026/94681334-28e48f80-031b-11eb-8dd5-010e6e23500c.png)

# Dynamic Rendering [![Gem Version](https://badge.fury.io/rb/rails-dynamic-rendering.svg)](https://badge.fury.io/rb/rails-dynamic-rendering) ![CI Badge](https://github.com/tricycle/rails-dynamic-rendering/workflows/RSpec%20Test%20Suite/badge.svg)

Is your SEO performance being held hostage by frontend JavaScript UI frameworks? Dynamic rendering is a "workaround solution" (in the words of Google) to help make difficult to index JavaScript based content easier for search engine crawlers to understand by presenting them with a pre-rendered HTML snapshot of your page.

This gem implements dynamic rendering for Rails applications using [Puppeteer](https://github.com/puppeteer/puppeteer) via the [Grover](https://github.com/Studiosity/grover) gem.

You can find out more about dynamic rendering in Google's article:
https://developers.google.com/search/docs/guides/dynamic-rendering

## Is this server-side rendering / SSR?

Kind of - Server-side rendering typically involves pre-rendering JavaScript elements of the page on the server within some isolated NodeJS environment and then having the client "re-hydrate" the DOM after loading. Dynamic rendering differs in that it's targeted specifically at crawler user agents and doesn't involve any "re-hydration" of the DOM, rather it's a static representation of the DOM, a snapshot of how the page looked in a headless Chrome instance.

## How does this work?

![diagram](https://user-images.githubusercontent.com/2643026/94683948-39970480-031f-11eb-9ea7-e90a03b0529b.jpg)

## Install

`gem 'rails-dynamic-rendering'`

## Usage

Include the concern and then use `enable_dynamic_rendering`. Any controller action covered by the `enable_dynamic_rendering` call (which accepts all the same arguments as `after_action`) will based on the user agent choose to either just return the HTML with JS or choose to render the page in a headless Chrome instance & render back the serialized HTML with all JS removed.

```ruby
class ApplicationController
  include DynamicRendering::ActsAsDynamicallyRenderable

  enable_dynamic_rendering only: :index, if: -> { some_condition_is_met? }

  def index
    @my_code = 'Does fancy stuff'
  end

  private

  def some_condition_is_met?
    true
  end
end

```

You can override the built in decision making thats used to determine whether a snapshot of the page should be returned by defining your own `#request_suitable_for_dynamic_rendering?` method. You might want to do this if you'd like to target other user agents:

```ruby
def request_suitable_for_dynamic_rendering?
  request.user_agent.include? 'My Problematic Bot That Does Not Render JS Content'
end
```


CHANGELOG.md
1.0.3

* Pass user-agent through to Puppeteer so that async requests are identified as being made from the original requestor.

1.0.2

* Switch user-agent check from String#match to Regexp#match

1.0.1

* Bumped Rake dependency to silence CVE-2020-8130 warning

1.0.0

* Initial release


spec/rails_helper.rb
# frozen_string_literal: true

ENV['RAILS_ENV'] ||= 'test'

require 'spec_helper'
require_relative "../spec/dummy/config/environment"
ActiveRecord::Migrator.migrations_paths = [File.expand_path("../test/dummy/db/migrate", __dir__)]
require 'rspec/rails'
require 'rails-dynamic-rendering'


spec/dummy/config/application.rb
require_relative 'boot'

require 'rails/all'

Bundler.require(*Rails.groups)
require "rails-dynamic-rendering"

module Dummy
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 6.0

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration can go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded after loading
    # the framework and any gems in your application.
  end
end


spec/lib/dynamic_rendering/acts_as_dynamically_renderable_spec.rb
# typed: false
# frozen_string_literal: true

require 'rails_helper'

module DynamicRendering
  describe ActsAsDynamicallyRenderable, type: :controller do
    described_class.tap do |concern|
      controller(ActionController::Base) do
        include concern

        enable_dynamic_rendering

        def index
          case params[:scenario]
          when 'ok-json'
            render json: { hello: 'world' }
          when 'not-ok-html'
            html = <<~HTML
              <html>
                <head><title>404!</title></head>
                <body>404!</body>
              </html>
            HTML
            render status: :not_found, html: html.html_safe
          when 'ok-html'
            html = <<~HTML
              <html>
                <head>
                  <title>Hello World</title>
                </head>
                <body>
                  <script type="text/javascript">
                    console.log('Hello world!');
                  </script>
                  <p>Hello world!</p>
                  <script type="application/ld+json">
                    { "hello": "world" }
                  </script>
                  <script>
                    var myDynamicNode = document.createElement("P");
                    myDynamicNode.appendChild(document.createTextNode("Hello Dynamic World"));
                    document.body.appendChild(myDynamicNode);
                  </script>
                </body>
              </html>
            HTML
            render html: html.html_safe
          else
            raise ArgumentError, "Unknown test scenario"
          end
        end
      end
    end

    describe '#render_dynamically' do
      subject(:make_request) do
        request.headers.merge!(headers)
        get :index, params: params
      end

      let(:params) do
        { scenario: scenario }
      end
      let(:headers) do
        { 'HTTP_USER_AGENT' => user_agent }
      end
      let(:scenario) { 'ok-html' }
      let(:user_agent) { 'Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1' }

      shared_examples_for 'it does not use dynamic rendering' do |expected_body, expected_status|
        specify do
          expect(DynamicRendering::ActsAsDynamicallyRenderable::HtmlProcessor)
            .not_to receive(:new)
          make_request
          expect(response.body).to match(expected_body)
          expect(response).to have_http_status(expected_status)
        end
      end

      context 'when the request is not coming from a supported crawler' do
        let(:user_agent) { 'Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1' }

        it_behaves_like 'it does not use dynamic rendering', /console.log/, :ok
      end

      context 'when the request is coming from a supported crawler' do
        let(:user_agent) { 'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)' }

        context 'when the response status is not ok' do
          let(:scenario) { 'not-ok-html' }

          it_behaves_like 'it does not use dynamic rendering', /404/, :not_found
        end

        context 'when the response status is ok' do
          context 'when the response type is not HTML' do
            let(:scenario) { 'ok-json' }

            it_behaves_like 'it does not use dynamic rendering', '{"hello":"world"}', :ok
          end

          context 'when the response type is HTML' do
            let(:html_processor) do
              instance_double(DynamicRendering::ActsAsDynamicallyRenderable::HtmlProcessor)
            end
            let(:expected_preprocessed_html) do
              <<~HTML
              <html>
                <head><script type="text/javascript">window.dynamicRendering = true;</script>
                  <title>Hello World</title>
                </head>
                <body>
                  <script type="text/javascript">
                    console.log('Hello world!');
                  </script>
                  <p>Hello world!</p>
                  <script type="application/ld+json">
                    { "hello": "world" }
                  </script>
                  <script>
                    var myDynamicNode = document.createElement("P");
                    myDynamicNode.appendChild(document.createTextNode("Hello Dynamic World"));
                    document.body.appendChild(myDynamicNode);
                  </script>
                </body>
              </html>
              HTML
            end
            let(:rendered_html) do
              <<~HTML
                <html>
                  <head><script type="text/javascript">window.dynamicRendering = true;</script>
                    <title>Hello World</title>
                  </head>
                  <body>
                    <script type="text/javascript">
                      console.log('Hello world!');
                    </script>
                    <p>Hello world!</p>
                    <script type="application/ld+json">
                      { "hello": "world" }
                    </script>
                    <script>
                      var myDynamicNode = document.createElement("P");
                      myDynamicNode.appendChild(document.createTextNode("Hello Dynamic World"));
                      document.body.appendChild(myDynamicNode);
                    </script>
                    <p>Hello Dynamic World</p>
                  </body>
                </html>
              HTML
            end

            before do
              expect(DynamicRendering::ActsAsDynamicallyRenderable::HtmlProcessor)
                .to receive(:new)
                .and_return(html_processor)
              expect(html_processor)
                .to receive(:convert)
                .with(
                  'content',
                  expected_preprocessed_html,
                  {
                    'displayUrl' => 'http://test.host/anonymous?scenario=ok-html',
                    'waitUntil' => 'networkidle2',
                    viewport: { width: 1400, height: 950 },
                    userAgent: instance_of(String)
                  }
                )
                .and_return(rendered_html)
              make_request
            end

            it 'removes javascript' do
              expect(response.body).not_to include('document.createTextNode')
            end

            it 'leaves JSON-LD in place' do
              expect(response.body).to include('{ "hello": "world" }')
            end
          end
        end
      end
    end
  end
end


lib/dynamic_rendering/acts_as_dynamically_renderable.rb
# typed: false
# frozen_string_literal: true

module DynamicRendering
  module ActsAsDynamicallyRenderable
    extend ActiveSupport::Concern

    USE_DYNAMIC_RENDERING_FOR_USER_AGENTS = /(googlebot|google-structured-data-testing-tool|bingbot)/i
    MOBILE_USER_AGENT = /mobile/i
    MOBILE_VIEWPORT = { width: 410, height: 730 }
    DESKTOP_VIEWPORT = { width: 1400, height: 950 }

    included do
      def render_dynamically(log_level: :info)
        return unless request_suitable_for_dynamic_rendering? && response_suitable_for_dynamic_rendering?

        Rails.logger.public_send(
          log_level,
          <<~TEXT
            [Dynamic rendering 🔍 ]:
            • URL: #{request.original_url}
            • User-Agent: #{request.user_agent}
            • Viewport: #{dynamic_rendering_viewport.inspect}
          TEXT
        )

        response.body = HtmlRenderer.new(
          response.body,
          dynamic_rendering_viewport,
          request.original_url,
          request.user_agent
        ).to_s
      end
    end

    class_methods do
      def enable_dynamic_rendering(arguments = {})
        append_after_action(:render_dynamically, **arguments)
      end
    end

    private

    def request_suitable_for_dynamic_rendering?
      USE_DYNAMIC_RENDERING_FOR_USER_AGENTS.match?(request.user_agent)
    end

    def dynamic_rendering_viewport
      return MOBILE_VIEWPORT if dynamic_rendering_request_from_mobile_crawler?

      DESKTOP_VIEWPORT
    end

    def dynamic_rendering_request_from_mobile_crawler?
      MOBILE_USER_AGENT.match?(request.user_agent)
    end

    def response_suitable_for_dynamic_rendering?
      response.ok? &&
        Mime::Type.lookup(response.media_type).html?
    end

    class HtmlRenderer
      DEFAULT_OPTIONS = { 'waitUntil' => 'networkidle2' }

      def initialize(original_response_body, viewport, original_url, original_user_agent, options = DEFAULT_OPTIONS)
        @original_response_body = original_response_body
        @viewport = viewport
        @original_url = original_url
        @options = options
        @original_user_agent = original_user_agent
      end

      def to_s
        HtmlPostProcessor.new(rendered_html).to_s
      end

      private

      def rendered_html
        processor.convert(
          'content',
          response_body_for_processor,
          @options.merge(
            'displayUrl' => @original_url,
            viewport: @viewport,
            userAgent: @original_user_agent
          )
        )
      end

      def processor
        HtmlProcessor.new(Dir.pwd)
      end

      def response_body_for_processor
        HtmlPreprocessor.new(@original_response_body).to_s
      end
    end

    class HtmlPreprocessor
      def initialize(html_as_string)
        @html_as_string = html_as_string
      end

      def to_s
        append_dynamic_rendering_variable!
        @html_as_string
      end

      private

      PRERENDER_VARIABLE = "<script type=\"text/javascript\">window.dynamicRendering = true;</script>"

      def append_dynamic_rendering_variable!
        @html_as_string.sub!(/<head[^>]*>/, "\\0#{PRERENDER_VARIABLE}")
      end
    end

    class HtmlPostProcessor
      def initialize(html_as_string)
        @html = Nokogiri::HTML(html_as_string)
      end

      def to_s
        remove_javascript!
        @html.to_s
      end

      private

      JAVASCRIPT_SELECTOR = 'script:not([type]), script[type="text/javascript"]'

      def remove_javascript!
        @html.css(JAVASCRIPT_SELECTOR).remove
      end
    end

    class HtmlProcessor < ::Grover::Processor
      def convert(method, url_or_html, options)
        spawn_process
        ensure_packages_are_initiated
        result = call_js_method method, url_or_html, options
        return unless result

        result
      ensure
        cleanup_process if stdin
      end
    end
  end
end
